File: TFMVVM-Example/TFMVVM-Example/View/SimpleHeaderViewWithFolding.m
//
//  SimpleHeaderFooterWithFolding.m
//  TFMVVM-Example
//
//  Created by Krzysztof on 08/08/2015.
//  Copyright (c) 2015 Trifork. All rights reserved.
//

#import "SimpleHeaderViewWithFolding.h"
#import "SimpleHeaderViewModel.h"
#import "TFSectionViewModel.h"

@interface SimpleHeaderViewWithFolding()
@property (nonatomic, strong) UILabel * primaryLabel;
@property (nonatomic, strong) UITapGestureRecognizer * tgr;
@property (nonatomic, strong) TFSectionViewModel * observedSection;
@end

@implementation SimpleHeaderViewWithFolding

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (!self) return nil;
    
    [self setup];
    return self;
}

- (instancetype)initWithReuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithReuseIdentifier:reuseIdentifier];
    if (!self) return nil;
    
    [self setup];
    return self;
}

- (void)setup
{
    self.contentView.backgroundColor = [UIColor blackColor];
    
    // Label
    _primaryLabel = [[UILabel alloc] init];
    _primaryLabel.font = [UIFont fontWithName:@"HelveticaNeue-Medium" size:16];
    _primaryLabel.textColor = [UIColor whiteColor];
    _primaryLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.contentView addSubview:_primaryLabel];
    
    self.contentView.frame = CGRectMake(0, 0, 100, 44); // in order to don't get layout warnings
    
    NSDictionary * dic = NSDictionaryOfVariableBindings(_primaryLabel);
    [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[_primaryLabel]|" options:0 metrics:nil views:dic]];
    [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[_primaryLabel]|" options:0 metrics:nil views:dic]];
}

- (void)configureWith:(SimpleHeaderViewModel *)headerViewModel
{
    NSParameterAssert([headerViewModel conformsToProtocol:@protocol(TFSectionItemViewModel)]);
    
    if (self.tgr) {
        [self removeGestureRecognizer:self.tgr];
    }
    
    _tgr = [[UITapGestureRecognizer alloc] initWithTarget:headerViewModel.sectionViewModel action:@selector(toggleFolding:)];
    [self addGestureRecognizer:_tgr];
 
    if (self.observedSection) {
        [self.observedSection removeObserver:self forKeyPath:@"folded" context:NULL];
    }
    self.observedSection = headerViewModel.sectionViewModel;
    [self.observedSection addObserver:self forKeyPath:@"folded" options:NSKeyValueObservingOptionNew context:NULL];
    
    [self setupText:headerViewModel.sectionViewModel];
}

- (void)setupText:(TFSectionViewModel *)vm
{
    [UIView transitionWithView:self.primaryLabel duration:0.4 options:UIViewAnimationOptionTransitionCrossDissolve animations:^{
        self.primaryLabel.text = vm.isFolded ?
            [NSString stringWithFormat:@"Closed with %lu elements, click to unfold", (unsigned long)[vm rows].count] :
            ((SimpleHeaderViewModel*)vm.header).title;
    } completion:nil];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if (object != self.observedSection) return;
    if (![keyPath isEqualToString:@"folded"]) return;
    
    [self setupText:object];
}

@end

File: TFMVVM-Example/TFMVVM-Example/View/SimpleHeaderViewWithFolding.swift
//
//  SimpleHeaderViewWithFolding.swift
//  TFMVVM-Example
//
//  Created by Krzysztof Profic on 11.01.2016.
//  Copyright © 2016 Trifork. All rights reserved.
//

import UIKit

// TODO rename "withSelection
class SimpleHeaderViewWithFolding: UITableViewHeaderFooterView, TFConfiguring {
    var primaryLabel: UILabel?
    var tgr: UITapGestureRecognizer?
    var observedSection: TFSectionViewModel?
    var removeButton: UIButton?
    
    override init(reuseIdentifier: String?) {
        super.init(reuseIdentifier: reuseIdentifier)
        self.commonInit()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.commonInit()
    }
    
    func commonInit(){
        tgr = UITapGestureRecognizer(target: nil, action: nil)
        self.addGestureRecognizer(tgr!)
        
        self.contentView.backgroundColor = UIColor.blackColor()
        primaryLabel = UILabel()
        primaryLabel?.font = UIFont(name: "HelveticaNeue-Medium", size: 16);
        primaryLabel?.textColor = UIColor.whiteColor()
        primaryLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.contentView.addSubview(primaryLabel!)
        
        removeButton = UIButton(type: .Custom)
        removeButton?.setTitle("[Remove]", forState: .Normal)
        removeButton?.translatesAutoresizingMaskIntoConstraints = false
        removeButton?.sizeToFit()
        self.contentView.addSubview(removeButton!)
        
        self.contentView.frame = CGRectMake(0, 0, 100, 44)
        
        let dic = ["primaryLabel": primaryLabel!, "removeButton": removeButton!]
        self.contentView.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:|[primaryLabel]|", options:.AlignAllBaseline, metrics:nil, views:dic))
        self.contentView.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:|[removeButton]|", options:.AlignAllBaseline, metrics:nil, views:dic))
        self.contentView.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:|[primaryLabel]-[removeButton]|", options:.AlignAllBaseline, metrics:nil, views:dic))
    }

    deinit {
        self.tgr?.removeTarget(nil, action: nil)
        self.removeButton?.removeTarget(nil, action: nil, forControlEvents: .TouchUpInside)
        self.observedSection?.removeObserver(self, forKeyPath: "folded", context: nil)
    }
    
    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        guard object as? TFSectionViewModel == self.observedSection else { return }
        guard keyPath == "folded" else { return }
        
        self.setupText(self.observedSection!)
    }
    
    // MARK: TFConfiguring
    func configureWith(object: AnyObject!) {
        guard let vm = object as? SimpleHeaderViewModel else { return }
        
        if (self.observedSection != nil) {
            self.observedSection?.removeObserver(self, forKeyPath: "folded", context: nil)
        }
        
        self.observedSection = vm.sectionViewModel
        self.observedSection?.addObserver(self, forKeyPath: "folded", options: .New, context: nil)
        self.setupText(vm.sectionViewModel!)
        
        // folded
        self.tgr?.removeTarget(nil, action: nil)
        self.tgr?.addTarget(vm.sectionViewModel!, action: "toggleFolding:")
        
        // removing
        self.removeButton?.removeTarget(nil, action: nil, forControlEvents: .TouchUpInside)
        self.removeButton?.addTarget(vm.sectionViewModel, action: "delete:", forControlEvents: .TouchUpInside)
    }
    
    private func setupText(vm: TFSectionViewModel) {
        UIView.transitionWithView(self.primaryLabel!, duration:0.4, options:.TransitionCrossDissolve, animations:{
            self.primaryLabel!.text = vm.folded ? "Closed with \(vm.rows.count) elements, click to unfold" : (vm.header as! SimpleHeaderViewModel).title;
        }, completion:nil)
    }
}
